### ISSUE TITLE
interfaces: multi-dhcp6c and multi-prefix delegation options

### ISSUE THREAD

**Issue by BazlTech** [opnsense/core#7647]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [X] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [X] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Is your feature request related to a problem? Please describe.**

For the purposes of IPv6 and pulling Ipv6 prefixes from my ISP (AT&T), I need a solution in which OPNSense makes the PD requests so that the Residential Gateway (AT&T RG "black box" device) will actually "activate" my available prefixes.  No matter what settings I make on the DHCPv6 client config for the WAN (under basic), the RG will at most only activate a single IPv6 prefix.

Currently, I have to use a custom DHCPv6 client config override file that makes those PD requests so the prefixes get "activated" on the RG device.  I only request 8 prefixes, because even though AT&T will give you a /60, it reserves 8 of them for itself.  Customers only get to use 8 of the 16 prefixes.

```
interface ixl0 {
    send ia-na 0;
    send ia-pd 0;
    send ia-pd 1;
    send ia-pd 2;
    send ia-pd 3;
    send ia-pd 4;
    send ia-pd 5;
    send ia-pd 6;
    send ia-pd 7;
    #request domain-name-servers;
    #request domain-name;
    script "/var/etc/dhcp6c_wan_script.sh";
};
id-assoc na 0 { };
id-assoc pd 0 {
    prefix-interface ixl3 {
        sla-id 0;
        sla-len 0;
    };
};
id-assoc pd 1 {
    prefix-interface ixl3 {
        sla-id 0;
        sla-len 0;
    };
};
id-assoc pd 2 {
    prefix-interface ixl3 {
        sla-id 0;
        sla-len 0;
    };
};
.
.
.
id-assoc pd 7 {
    prefix-interface ixl3 {
        sla-id 0;
        sla-len 0;
    };
};
```

**Describe the solution you like**

It would be nice if there was something like the above already built in, and I could simply tick the boxes/make drop-down selections on what interfaces to make the PD selections to, and how many.  In my case, they all go to the same interface, since I'm not using OPNSense to manage VLANs.  I just need the prefixes made available on the RG.  If you were using OPNSense to manage all of your VLANs, then each "id-assoc pd #' line would relate to that specific VLAN interface.

**Describe alternatives you considered**

Using this custom dhcpv6 client config override file is the only thing I've found that works.

**Comment by fichtner** [comment]
@BazlTech Thanks for the ticket. I'll tentatively take this as discussed and be back with a couple of questions when the dust on 24.7 has settled. :)

**Comment by BazlTech** [comment]
> @BazlTech Thanks for the ticket. I'll tentatively take this as discussed and be back with a couple of questions when the dust on 24.7 has settled. :)

Any idea if this might be worked into an upcoming release?

**Comment by BazlTech** [comment]
Just checking on an update.

**Comment by BazlTech** [comment]
Is this still slated for 25.7?  The RC doesn't look like it includes these settings for WAN, only for kea.

**Comment by sopex** [comment]
> Is this still slated for 25.7? The RC doesn't look like it includes these settings for WAN, only for kea.

Very unlikely, unfortunately, I don't see any code mentioning this issue :(

**Comment by fichtner** [comment]
@BazlTech I don't have enough time in community scope at the moment to get this closer to being usable. I'm moving it to 26.1 for reeval.

**Comment by BazlTech** [comment]
> [@BazlTech](https://github.com/BazlTech) I don't have enough time in community scope at the moment to get this closer to being usable. I'm moving it to 26.1 for reeval.

Thanks for the update!  Doing awesome work!

**Comment by fichtner** [comment]
@BazlTech I've made a small effort towards this with 3e810bd4be538. I still need to check a few things, but if I understand correctly this is just a configuration thing and not related to any work that is required inside dhcp6c.


Cheers,
Franco

**Comment by fichtner** [comment]
PS: do you need a way to disable DNS options too? Judging by the config you disabled it.

**Comment by BazlTech** [comment]
> [@BazlTech](https://github.com/BazlTech) I've made a small effort towards this with [3e810bd](https://github.com/opnsense/core/commit/3e810bd4be538903259a46cd57b1118269080740). I still need to check a few things, but if I understand correctly this is just a configuration thing and not related to any work that is required inside dhcp6c.
> 
> Cheers, Franco

Right.  It’s just a PD config request that allows me to tell the AT&T RG that it needs to request the extra prefixes.  I won’t have opnSense be serving out DHCPv6 IPv6 leases, but I need it to make that PD request in a very specific manner to the RG so those IPv6 prefixes get “activated.”

And correct on the DNS setting - it would be good to include that option to enable/disable as needed.

**Comment by fichtner** [comment]
@BazlTech Clear, thanks. I'll try to provide a test patch in the second week of January after I can test this in my main setup.


Cheers,
Franco

**Comment by BazlTech** [comment]
> [@BazlTech](https://github.com/BazlTech) Clear, thanks. I'll try to provide a test patch in the second week of January after I can test this in my main setup.
> 
> Cheers, Franco

Great, thanks or the update!

**Comment by fichtner** [comment]
@BazlTech this is turning into a long-term project with overlap in the development branch due to other shortcomings and weirdness.. I'm not sure anything testable will be available this week. FWIW, d4e0109405cfe232c has the code restructured so that each LAN (well, for now only WAN, but the idea is the same) will be able to select its own `id-assoc pd` ID and the config will generate the relevant `send ia-pd`. DNS can be turned off too. It also splits the dhcp6c configuration into multiple daemons for each WAN because the current automatic `$id` handling is rather wrong when a second WAN comes in.


Cheers,
Franco

**Comment by fichtner** [comment]
@BazlTech the development version of the upcoming 25.7.11 + https://github.com/opnsense/core/commit/78198c0478 should do what you want. You can disable requesting DNS on WAN and each LAN has a "Optional IA-PD ID" you can set (zero is not accepted but is the default anyway).

I'm unsure about `sla-id` and `sla-len` now that I'm looking at your config again. `sla-len` should be fine but `sla-id` may use a wrong validation. Should be easy to fix. Let me check. ;)

**Comment by fichtner** [comment]
Ok 25.7.11 development + 12e3d0b79632 is probably how it should go :)

**Comment by fichtner** [comment]
I'll do a call for testing in the forum. The milestone moves to 26.7 but I'm sure we can get this into 26.1.x quickly. Just need some testers with multi-WAN DHCPv6.


Cheers,
Franco

**Comment by fichtner** [comment]
Closing this as code landed on master finally, but will do another call for testing based on 26.1.3 in a week or two.

**Comment by fichtner** [comment]
This never landed in 26.1.x but I've worked on this today to bring this into 26.1.9 in a few weeks without the multi-dhcp6c support which didn't work out too well in the public CFT.

**Comment by BazlTech** [comment]
> This never landed in 26.1.x but I've worked on this today to bring this into 26.1.9 in a few weeks without the multi-dhcp6c support which didn't work out too well in the public CFT.

Thanks for the update!

**Comment by fichtner** [comment]
@BazlTech so as promised the backport https://github.com/opnsense/core/commit/91093f334 adds the IAID selection for each prefix ID you can track.. also includes an unrelated change for reserving a prefix range for a different ticket.

    # opnsense-patch 91093f334

This is specifically for 26.1.8. Let me know what you think.


Cheers,
Franco

**Comment by fichtner** [comment]
@maurice-w maybe you also want to try https://github.com/opnsense/core/commit/91093f334 at least to see if the IAID defaults are ok with this patch as previously discussed
